src/main/java/com/rackspace/jenkins_nodepool/Janitor.java

```diff
@@ -160,40 +181,104 @@ void cleanNode(NodePoolSlave nodePoolSlave, String reason) {
+                                model.getAllocated_to()));
+                    } else if (run != null && model.getBuild_id() != null && run.isBuilding()) {
+                        // Node is being used: we have an external id, a build id, and it's currently building
+                        LOG.log(FINE, format("In-Use Node: %s, region: %s, held: %b, allocated to: %s, build id: %s, is running: %s",
+                                nodePath,
+                                model.getRegion(),
+                                (model.getHold_job() != null),
+                                model.getAllocated_to(),
+                                model.getBuild_id(),
+                                run.isBuilding()));
+                    } else if (run != null && model.getBuild_id() != null && !run.isBuilding() && model.getHold_job() != null) {
+                        LOG.log(FINE, format("Retired Node (held): %s, region: %s, held: %b, allocated to: %s, build id: %s, is running: %s.",
+                                nodePath,
+                                model.getRegion(),
+                                (model.getHold_job() != null),
+                                model.getAllocated_to(),
+                                model.getBuild_id(),
+                                run.isBuilding()));
+                    } else if (run != null && model.getBuild_id() != null && !run.isBuilding() && model.getHold_job() == null) {
+                        LOG.log(FINE, format("Retired Node (not held): %s, region: %s, held: %b, allocated to: %s, build id: %s, is running: %s - should REMOVE this node.",
+                                nodePath,
+                                model.getRegion(),
+                                (model.getHold_job() != null),
+                                model.getAllocated_to(),
+                                model.getBuild_id(),
+                                run.isBuilding()));
+
+                        // Toggle the state to used
+                        model.setState(NodePoolState.USED.getStateString());
+                        // Save the changes back to ZK
+                        final String jsonStringModel = gson.toJson(model, ZKNodeModel.class);
+                        conn.setData().forPath(path, jsonStringModel.getBytes(StandardCharsets.UTF_8));
+                        LOG.log(FINE, format("Retired Node (not held): %s, region: %s, held: %b, allocated to: %s, build id: %s, is running: %s - changed state to USED - launcher should remove.",
+                                nodePath,
+                                model.getRegion(),
+                                (model.getHold_job() != null),
+                                model.getAllocated_to(),
+                                model.getBuild_id(),
+                                run.isBuilding()));
+                        // TODO: DAD - Is this enough to trigger the NP Launcher to clean up the node??
```

Comment:
You probably need to release a lock if it exists https://github.com/jenkinsci/nodepool-agents-plugin/blob/226ad8bd2bb9c44ab670eb9bc6ed2ab33b63fb3a/src/main/java/com/rackspace/jenkins_nodepool/NodePoolNode.java#L281-L287

I know I've made the point above but if you incorporated your model into NPN, then you'd have the function for releasing, and could use the code in ZooKeeperObject for interacting with ZK.  It may require some refactoring to turn NPN into a model class, so you can create one without a NPJ reference. Shouldn't be too bad though, override the constructor and null-check NPJ accesses? OR... get the real NPJ object as you already know the build id, and use that to construct the NPN. Might need to add a function to NPJH to get job by externalizable id. #ThinkingInAcomment